.aspect/external_repository_hashes/npm_translate_lock_NDg3NzUwNzA2NQ==

```diff
@@ -0,0 +1,16 @@
+# Input hashes for repository rule: npm_translate_lock(name = "npm", pnpm_lock = "@//:pnpm-lock.yaml")
```

Comment:
I still need help to understand why this file is needed. What goes wrong if we just use those files as inputs to the repository rule directly? Does it sometimes invalidate when it shouldn't? Fail to invalidate when it should?

Reply:
This allows the repo rule to update the `pnpm-lock.yaml` file in the source tree when its inputs change but only for the developer who makes the changes to the inputs. Another developer who pulls the changes will invalidate the repo rule but it will determine that the `pnpm-lock.yaml` file is up-to-date with its inputs so it won't run `pnpm install --lockfile-only` or `pnpm import` (depending on your SoT)

Reply:
This is also better than build action since developer machines typically don't share a remote cache with CI or with other developers; especially for developers on macos. Making the lock file update a build action would mean that every developer in a monorepo would have to have pnpm fetch 100% of the deps in the monorepo even if they don't build any js targets. `pnpm install` and `pnpm import` will cache the npm packages but that cost will always have to be paid to bootstrap. With the external repository hash, only javascript developers in a monorepo will have to ever fetch npm packages to their machine with pnpm.

Reply:
Perhaps `.aspect/external_repository_action_cache/` is clearer naming

Reply:
Yeah I get it. You're basically creating something like an action graph inside repository rules. The first one is expected to invalidate frequently but is quick to run, and as long as it produces a stable output it prevents the next "action" from running.

If you're going to check in a file anyway, could it also contain the translated pnpm lock? Just make it a .bzl file that does both? That would make this more similar to how Java, Go and Python work when you vendor the "pinned" file as starlark. The translated file doesn't have to be portable to work as an external repo, like we ran into with rules_ts.

Does that make sense?

Reply:
Yeah. I think I know what you're getting at.

Reply:
The pnpm lock file is translated to a number of starlark files. There is the `npm_link_all_packages()` function that is the primary one but there are also the `package_json.bzl` pass-through stubs & the root BUILD file. Wouldn't make any sense to vendor just part of this generation since you'd still have to run the same generation code for the rest. I was thinking of vendoring in some intermediate format that is post calculating the transitive closure but that would require a lot more refactoring than already done here. So just checking the hashes of the inputs as is already done here is probably best for this step function forward IMO.

Reply:
yup, that makes sense